File: exflag.py
# =====================================================
#          exflag1 (binary oracle), py edition
#                     by Arc'blroth
# =====================================================

from typing import Callable, List


def parse_glob(glob: str) -> List[int]:
    """
    "parse" the charset "glob"
    """
    lexed_glob = []
    parse_pos = 0
    eof_pos = len(glob)
    while parse_pos < len(glob):
        if glob[parse_pos] == "\\":
            if parse_pos + 1 == eof_pos:
                raise SyntaxError('truncated escape in charset "glob"')
            match glob[parse_pos + 1]:
                case "\\":
                    lexed_glob.append("\\")
                case "-":
                    lexed_glob.append("-")
                case _:
                    raise SyntaxError('invalid escape in charset "glob"')
            parse_pos += 2
        elif glob[parse_pos] == "-":
            lexed_glob.append(None)
            parse_pos += 1
        else:
            lexed_glob.append(glob[parse_pos])
            parse_pos += 1
    parsed_glob = []
    parse_pos = 0
    eof_pos = len(lexed_glob)
    while parse_pos < len(lexed_glob):
        if lexed_glob[parse_pos] is None:
            raise SyntaxError('invalid range in charset "glob"')
        if parse_pos + 1 != eof_pos and lexed_glob[parse_pos + 1] is None:
            if parse_pos + 2 == eof_pos:
                raise SyntaxError('invalid range in charset "glob"')
            parsed_glob += list(range(ord(lexed_glob[parse_pos]), ord(lexed_glob[parse_pos + 2]) + 1))
            parse_pos += 3
        else:
            parsed_glob.append(ord(lexed_glob[parse_pos]))
            parse_pos += 1
    return parsed_glob


def extract_flag(
    oracle: Callable[[str], bool],
    prefix: str = "",
    charset: List[int] = "_0-9A-Za-z}",
    stop: Callable[[str], bool] = lambda x: x.endswith("}"),
) -> str:
    """
    Exfiltrates a flag, character by character, given a binary oracle and a charset.

    :param oracle: a binary oracle that checks if the given prefix of the flag is valid
    :param prefix: known flag prefix to start with
    :param charset: a charset, in "glob"-like format, to try characters from
    :param stop: stopping oracle (defaults to flag having `}`)
    :raises Exception: if a flag cannot be found using the given charset
    """
    charset = parse_glob(charset)
    if not charset:
        raise ValueError("charset is empty")
    flag = prefix
    while not stop(flag):
        flag_len = len(flag)
        for char in charset:
            print(f'\x1b[90mtrying "{flag}{chr(char)}"\x1b[0m')
            if oracle(flag + chr(char)):
                flag += chr(char)
                print(f'\x1b[33mfound "{flag}"\x1b[0m')
                break
        if len(flag) == flag_len:
            raise Exception("failed to solve next flag character, aborting")

File: exflag.js
// =====================================================
//          exflag1 (binary oracle), js edition
//                     by Arc'blroth
// =====================================================

/**
* "parses" the charset "glob"
* 
* @param {string} glob 
* @returns {number[]}
*/
function parseGlob(glob) {
  let lexedGlob = []
  let parsePos = 0
  let eofPos = glob.length
  while (parsePos < glob.length) {
    if (glob[parsePos] === "\\") {
      if (parsePos + 1 === eofPos) throw "truncated escape in charset \"glob\""
      switch (glob[parsePos + 1]) {
      case "\\":
        lexedGlob.push("\\")
        break
      case "-":
        lexedGlob.push("-")
        break
      default:
        throw "invalid escape in charset \"glob\""
      }
      parsePos += 2
    } else if (glob[parsePos] === "-") {
      lexedGlob.push(null)
      parsePos++
    } else {
      lexedGlob.push(glob[parsePos])
      parsePos++
    }
  }
  let parsedGlob = []
  parsePos = 0
  eofPos = lexedGlob.length
  while (parsePos < lexedGlob.length) {
    if (lexedGlob[parsePos] === null) throw "invalid range in charset \"glob\""
    if ((parsePos + 1 !== eofPos) && (lexedGlob[parsePos + 1] === null)) {
      if (parsePos + 2 === eofPos) throw "invalid range in charset \"glob\""
      for(let i = lexedGlob[parsePos].charCodeAt(0); i <= lexedGlob[parsePos + 2].charCodeAt(0); i++) {
        parsedGlob.push(i)
      }
      parsePos += 3
    } else {
      parsedGlob.push(lexedGlob[parsePos].charCodeAt(0))
      parsePos += 1
    }
  }
  return parsedGlob
}

/**
* Exfiltrates a flag, character by character, given a binary oracle and a charset.
* 
* @param {(flag: string) => Promise<boolean>} oracle - a binary oracle that checks if the given prefix of the flag is valid
* @param {string} prefix - known flag prefix to start with
* @param {string} charset - a charset, in "glob"-like format, to try characters from
* @param {(flag: string) => boolean} stop - stopping oracle (defaults to flag having `}`)
* @param {(...args: any[]) => Promise<void>} log - log function (defaults to `console.log`)
* @returns {Promise<string>}
* @throws if flag could not be assembled with the given charset
*/
export default async function exflag(
  oracle,
  prefix="",
  charset="_0-9a-zA-Z",
  stop=f=>f.endsWith("}"),
  log=console.log,
) {
  charset = parseGlob(charset)
  if (!charset) throw "charset is empty"
  
  let flag = prefix
  while (!stop(flag)) {
    const flagLen = flag.length
    for (const char of charset) {
      const actualChar = String.fromCharCode(char)
      await log(`\x1b[90mtrying "${flag}${actualChar}"\x1b[0m`)
      if (await oracle(flag + actualChar)) {
        flag += actualChar
        await log(`\x1b[33mfound "${flag}"\x1b[0m`)
        break
      }
    }
    if (flag.length === flagLen) {
      throw "failed to solve next flag character, aborting"
    }
  }

  return flag
}
